tests/test_mixed_simulator_device.py

```diff
@@ -0,0 +1,910 @@
+
+    def test_sample_values(self, simulator_device_2_wires, tol):
+        """Tests if the samples returned by sample have
+        the correct values
+        """
+
+        simulator_device_2_wires.reset()
+
+        simulator_device_2_wires.apply([qml.RX(1.5708, wires=[0])])
+        simulator_device_2_wires._samples = simulator_device_2_wires.generate_samples()
+
+        s1 = simulator_device_2_wires.sample(qml.PauliZ(0))
+
+        # s1 should only contain 1 and -1, which is guaranteed if
+        # they square to 1
+        assert np.allclose(s1 ** 2, 1, **tol)
+
+
+class TestState:
+    """Test the state property."""
+
+    @pytest.mark.parametrize("shots,analytic", [(100, True)])
+    @pytest.mark.parametrize(
+        "ops,expected_pure_state",
+        [
+            ([qml.PauliX(0), qml.PauliX(1)], [0, 0, 0, 1]),
+            ([qml.PauliX(0), qml.PauliY(1)], [0, 0, 0, 1j]),
+            ([qml.PauliZ(0), qml.PauliZ(1)], [1, 0, 0, 0]),
+        ],
+    )
+    def test_state_pauli_operations(self, simulator_device_2_wires, ops, expected_pure_state, tol):
+        """Test that the state reflects Pauli operations correctly."""
+        simulator_device_2_wires.reset()
+        simulator_device_2_wires.apply(ops)
+
+        state = np.array(expected_pure_state)
+        expected_output = np.kron(state, state.conj()).reshape([4, 4])
+
+        assert np.allclose(simulator_device_2_wires.state, expected_output, **tol)
+
+    @pytest.mark.parametrize("shots,analytic", [(100, True)])
+    @pytest.mark.parametrize(
+        "ops,diag_ops,expected_pure_state",
+        [
+            ([qml.PauliX(0), qml.PauliX(1)], [], [0, 0, 0, 1]),
+            (
+                [qml.PauliX(0), qml.PauliY(1)],
+                [qml.Hadamard(0)],
+                [0, 1j / np.sqrt(2), 0, -1j / np.sqrt(2)],
+            ),
+            (
+                [qml.PauliZ(0), qml.PauliZ(1)],
+                [qml.Hadamard(1)],
+                [1 / np.sqrt(2), 1 / np.sqrt(2), 0, 0],
+            ),
+        ],
+    )
+    def test_state_pauli_operations_and_observables(
+        self, simulator_device_2_wires, ops, diag_ops, expected_pure_state, tol
+    ):
+        """Test that the state reflects Pauli operations and observable rotations correctly."""
+        simulator_device_2_wires.reset()
+        simulator_device_2_wires.apply(ops, rotations=diag_ops)
+
+        state = np.array(expected_pure_state)
+        expected_output = np.kron(state, state.conj()).reshape([4, 4])
+
+        assert np.allclose(simulator_device_2_wires.state, expected_output, **tol)
+
+    @pytest.mark.parametrize("shots,analytic", [(100, False)])
+    def test_state_non_analytic(self, simulator_device_2_wires):
+        """Test that the state is None if in non-analytic mode."""
+        simulator_device_2_wires.reset()
+        simulator_device_2_wires.apply([qml.PauliX(0), qml.PauliX(1)])
+
+        assert simulator_device_2_wires.state is None
```

Comment:
:100:

These are much more 'functional integration tests', and a prime candidate to be moved to a `pl-device-test` CLI utility. We'll have to make sure though that they can properly deal with both statevectors and density matrices, as well as plugins that insert global phase. Perhaps we'll just have to test that the output probability is correct.

Reply:
Most of the previous tests that tested the "state" actually tested the comp. basis probabilities.